File: GIMP/src/InOutFun/gidpost/examples/testpost_fd.f90
PROGRAM EXAMPLE
  use gidpost
  implicit none
  
  REAL(8) :: nodes(1:9,1:3), rx, ry, rz, value
  INTEGER :: elems(1:4,1:4)
  INTEGER :: idx, x, y, nid(1:4)
  CHARACTER(LEN=1024) :: nombre
  CHARACTER(LEN=1024) :: component_names(2)
  TYPE(GiD_File) :: fdm, fdr
  REAL(8) :: gpa

  idx = 1
  do x=0,2
    do y=0,2
      nodes(idx,1) = x
      nodes(idx,2) = y
      nodes(idx,3) = 0
      idx = idx + 1
    end do
  end do
  idx = 1
  do x=1,2
    do y=1,2
      elems(idx,1) = x + (y-1)*3
      elems(idx,2) = elems(idx,1) + 1
      elems(idx,3) = elems(idx,1) + 4
      elems(idx,4) = elems(idx,1) + 3
      idx = idx + 1
    end do
  end do

  CALL GiD_PostInit

  !CALL GID_OPENPOSTMESHFILE('testfortran90.post.msh',GiD_PostAscii)      
  !CALL GiD_OpenPostResultFile('testfortran90.post.res',GiD_PostBinary)      
  !CALL GiD_OpenPostResultFile('testfortran90.post.res',GiD_PostHDF5)  
  fdm=GiD_fOpenPostMeshFile('test_fdfortran90.post.msh',GiD_PostAscii)  
  fdr=GiD_fOpenPostResultFile('test_fdfortran90.post.res',GiD_PostAscii)
  
  !fdm=fdr
  
  CALL GiD_fBeginMeshColor(fdm,'quadmesh',GiD_2D,GiD_Quadrilateral,4,0.7d0,0.7d0,0.4d0)
  CALL GiD_fMeshUnit(fdm,'mm')
  CALL GiD_fBeginCoordinates(fdm)
  
  do idx=1,9         
    CALL GiD_fWriteCoordinates(fdm,idx,nodes(idx,1),nodes(idx,2),nodes(idx,3))      
  end do
  CALL GiD_fEndCoordinates(fdm)
  CALL GiD_fBeginElements(fdm)  
  do idx=1,4
    nid(1) = elems(idx,1)
    nid(2) = elems(idx,2)
    nid(3) = elems(idx,3)
    nid(4) = elems(idx,4)
    CALL GiD_fWriteElement(fdm,idx,nid)    
  end do      
  CALL GiD_fEndElements(fdm)
  CALL GiD_fEndMesh(fdm)  
  
  !Gauss points, declared but not used by any result  
  CALL GiD_fBeginGaussPoint(fdr,'element_gp',GiD_Quadrilateral,GiD_NULL,4,0,0)
  gpa = 0.57735027  
  CALL GiD_fWriteGaussPoint2D(fdr,-gpa,-gpa)
  CALL GiD_fWriteGaussPoint2D(fdr,gpa,-gpa)
  CALL GiD_fWriteGaussPoint2D(fdr,0.57735027d0,0.57735027d0)
  CALL GiD_fWriteGaussPoint2D(fdr,-gpa,gpa)
  CALL GiD_fEndGaussPoint(fdr)
  
  !Scalar result    
  CALL GiD_fBeginResultHeader(fdr,'Pressure','Analysis',1.0d0,GiD_Scalar,GiD_onNodes,GiD_NULL)  
  !CALL GiD_fScalarComp(fdr,'Comp')
  CALL GiD_fResultUnit(fdr,'Pa')
  CALL GiD_fResultValues(fdr)
  !CALL GiD_fBeginScalarResult(fdr,'Pressure','Analysis',1.0d0,GiD_onNodes,GiD_NULL,GiD_NULL,GiD_NULL) 
  do idx=1,9
    value=idx*1.5
    CALL GiD_fWriteScalar(fdr,idx,value)
  end do
  CALL GiD_fEndResult(fdr)
  CALL GiD_fFlushPostFile(fdr)
  !Vectorial result
  CALL GiD_fBeginResultHeader(fdr,'Velocity','Analysis',1.0d0,GiD_Vector,GiD_onNodes,GiD_NULL)
  CALL GiD_fVectorComp(fdr,'Comp X','Comp Y','Comp Z','Module')
  CALL GiD_fResultUnit(fdr,'m/s')
  CALL GiD_fResultValues(fdr)
  do idx=1,9
    value=idx*1.5
    CALL GiD_fWrite2DVector(fdr,idx,value,value*3)
  end do
  CALL GiD_fEndResult(fdr)

  !Complex Scalar result
  CALL GiD_fBeginResultHeader(fdr,'Potential','Analysis',1.0d0,GiD_ComplexScalar,GiD_onNodes,GiD_NULL)    
  component_names(1)='real_part'
  component_names(2)='imaginary_part'
  CALL GiD_fResultComponents(fdr,2,component_names)
  CALL GiD_fResultUnit(fdr,'Volt')
  CALL GiD_fResultValues(fdr)
  !CALL GiD_fBeginComplexScalarResult(fdr,'Complex scalar','Analysis',1.0d0,GiD_onNodes,GiD_NULL,GiD_NULL,'real_part','imaginary_part')
  do idx=1,9
    value=idx*1.5
    CALL GiD_fWriteComplexScalar(fdr,idx,value,value*1.2)
  end do
  CALL GiD_fEndResult(fdr)

  !CALL GiD_fClosePostMeshFile(fdm)
  CALL GiD_fClosePostResultFile(fdr)
  
  CALL GiD_PostDone
END PROGRAM

File: GraMS_V1/src/Future_Features/gidpost/examples/testpost_fd.c
#include <stdlib.h>
#include "gidpost.h"

/*
  #define ASCII_FORMAT
  #define DO_FLUSH
  #define EARLY_TERMINATE
*/

#define USE_GP_IN_GROUP

#if defined(USE_GP_IN_GROUP)
#define LPL 3
#define LOC elems
#else
#define LPL 1
#define LOC nodes
#endif

typedef struct {
  int id;
  float x, y, z;
} SCoord;

typedef struct {
  int id;
  int n1, n2, n3;
} SElem;

SCoord nodes[9];
SElem elems[8];

void GenMesh()
{
  int i, j, idx;
  SCoord * node;
  SElem * elm1, * elm2;

  idx = 1;
  for ( i = 0; i < 3; i++ ) {
    for ( j = 0; j < 3; j++, idx++ ) {
      node = nodes + idx -1;
      node->id = idx;
      node->x = (float)(i);
      node->y = (float)(j);
      node->z = 0;
    }
  }
  idx = 0;
  for ( i = 0; i < 2; i++ ) {
    for ( j = 0; j < 2; j++, idx+=2 ) {
      elm1 = elems + idx;
      elm2 = elm1 + 1;
      elm1->id = idx+1;
      elm1->n1 = i*3 + j + 1;
      elm1->n3 = i*3 + j + 1 + 3 ;
      elm1->n2 = elm1->n3 + 1;
      elm2->id = idx+2;
      elm2->n1 = elm1->n1;
      elm2->n2 = elm1->n1 + 1;
      elm2->n3 = elm1->n2;
    }
  }
}

float Random()
{
  return rand()/(float)(RAND_MAX);
}

int main()
{
  int i, j;
  SCoord * node;
  SElem * elm;
  int elmi[4];
  GiD_FILE fdm, fdr;

  GenMesh();

  GiD_PostInit();

#if defined(ASCII_FORMAT)
  fdm = GiD_fOpenPostMeshFile( "test_fd.post.msh", GiD_PostAscii);
  fdr = GiD_fOpenPostResultFile( "test_fd.post.res", GiD_PostAsciiZipped);
#else
  fdm = fdr = GiD_fOpenPostResultFile( "test_fd.post.res", GiD_PostBinary);
#endif

  /* write mesh info */
  GiD_fBeginMeshColor(fdm, "TestMsh", GiD_2D, GiD_Triangle, 3, 0, 0.99, 0);
  /* coordinates */
  GiD_fBeginCoordinates(fdm);
  for ( i = 0; i < 9; i++ ) {
    node = nodes + i;
    GiD_fWriteCoordinates(fdm, node->id, node->x, node->y, node->z ); 
  }
  GiD_fEndCoordinates(fdm);
  /* elements */
  GiD_fBeginElements(fdm);
  for ( i = 0; i < 8; i++ ) {
    elm = elems + i;
    elmi[0] = elm->n1;
    elmi[1] = elm->n2;
    elmi[2] = elm->n3;
    elmi[3] = 2;
    GiD_fWriteElementMat(fdm, elm->id, elmi);
  }
  GiD_fEndElements(fdm);
  GiD_fEndMesh(fdm);

  /* now results info */

  /*
    #define SIMPLE_RESULT
  */
#ifdef SIMPLE_RESULT
  GiD_fBeginResult(fdr,
                   "EscalarNodos", "Analysis", 1.0, GiD_Scalar, GiD_OnNodes,
		  NULL, NULL, 0, NULL);
  for ( i = 0; i < 9; i++ ) { 
    GiD_fWriteScalar(fdr, nodes[i].id, Random());
  }
  GiD_fEndResult(fdr);
#else
  /* write the gauss points */
  GiD_fBeginGaussPoint(fdr, "element_gp", GiD_Triangle, NULL, 3, 0, 1);  
  GiD_fEndGaussPoint(fdr);
  
  /* scalar result over nodes */
  /*
    #if defined(ASCII_FORMAT)  
  */
#if defined(USE_GP_IN_GROUP)
  GiD_fBeginResultGroup(fdr, "Analysis", 1.0, GiD_OnGaussPoints,"element_gp" );
#else
  GiD_fBeginResultGroup(fdr, "Analysis", 1.0, GiD_OnNodes, NULL);
#endif
  GiD_fResultDescription(fdr, "EscalarNodosG", GiD_Scalar);
  GiD_fResultDescriptionDim(fdr, "VectorNodosG", GiD_Vector, 4);
  GiD_fResultDescription(fdr, "MatrixG", GiD_Matrix);
  GiD_fResultDescription(fdr, "Local AxesG", GiD_LocalAxes);
  for ( i = 0; i < sizeof(LOC)/sizeof(LOC[0]); i++ ) {
    for (j = 0; j < LPL; j++) {
      GiD_fWriteScalar(fdr, LOC[i].id, Random());
      GiD_fWriteVectorModule(fdr, LOC[i].id, Random(), Random(), Random(),-1);
      GiD_fWrite3DMatrix(fdr, LOC[i].id, Random(), Random(), Random(),
                         Random(), Random(), Random());
      GiD_fWriteLocalAxes(fdr, LOC[i].id, Random(), Random(), Random());
    }
  }
  GiD_fEndResult(fdr);
  
  /* #else */

  /* scalar results over gauss points: Triangle + 3 GP */
  
  GiD_fBeginResult(fdr,
                   "SobreGPTriangle3", "Analysis", 1.0, GiD_Scalar,
                   GiD_OnGaussPoints, "element_gp", NULL, 0, NULL);
  for ( i = 0; i < 8; i++ ) {
    for ( j = 0; j < 3; j++ )
      GiD_fWriteScalar(fdr, i+1, Random());
  }
  GiD_fEndResult(fdr);
  GiD_fBeginResult(fdr,
                   "EscalarNodos", "Analysis", 1.0, GiD_Scalar, GiD_OnNodes,
		  NULL, NULL, 0, NULL);
  for ( i = 0; i < 9; i++ ) { 
    GiD_fWriteScalar(fdr, nodes[i].id, Random());
  }
  GiD_fEndResult(fdr);

  /* vector result over nodes */
  
  /*
  GiD_fBeginResult("VectorNodos", "Analysis", 1.0, GiD_Vector, GiD_OnNodes,
  NULL, NULL, 0, NULL);*/
  GiD_fBeginResultHeader(fdr,
                         "VectorNodos", "Analysis", 1.0, GiD_Vector, GiD_OnNodes, NULL);
  for ( i = 0; i < 9; i++ ) { 
    GiD_fWriteVector(fdr,
                     nodes[i].id, Random(), Random(), Random());
  }
  GiD_fEndResult(fdr);

#ifndef DISABLE_CODE

#ifdef DO_FLUSH
  GiD_fFlushPostFile(fdr);
#endif 
  /* matrix result */

  GiD_fBeginResult(fdr,
                   "Matrix", "Analysis", 1.0, GiD_Matrix, GiD_OnNodes,
                   NULL, NULL, 0, NULL);
  for ( i = 0; i < 9; i++ ) {
    GiD_fWrite3DMatrix(fdr, i+1, Random(), Random(), Random(),
                       Random(), Random(), Random());
  }
  GiD_fEndResult(fdr);  

#ifdef EARLY_TERMINATE
  abort();
#endif
  /* local axes result */

  GiD_fBeginResult(fdr,
                   "Local Axes", "Analysis", 1.0, GiD_LocalAxes, GiD_OnNodes,
                   NULL, NULL, 0, NULL);
  for ( i = 0; i < 9; i++ ) {
    GiD_fWriteLocalAxes(fdr, i+1, Random(), Random(), Random());
  }
  GiD_fEndResult(fdr);  
  /* #endif */ /* ASCII_FORMAT */
#endif /* DISABLE_CODE */
#endif /* SIMPLE_RESULT */
  
#ifdef ASCII_FORMAT
  GiD_fClosePostMeshFile(fdm);
#endif
  GiD_fClosePostResultFile(fdr);

  GiD_PostDone();

  return 0;
}
